tripy/tripy/frontend/utils.py

```diff
@@ -245,26 +245,44 @@ def wrapper(*args, **kwargs):
-            # because ShapeScalars will typically be broadcasted in order to operate
-            # with Tensors/Shapes. Otherwise, Shape and ShapeScalar would have been
-            # at the same level of hierarchy.
-            #
-            # When a function has multiple arguments, the order of precedence is:
-            #
-            # 1. Tensor
-            # 2. Shape
-            # 3. ShapeScalar
-            #
-            # That is, if we have an operation with a mixture of types, all arguments are
-            # converted to the type with the highest precedence, e.g. Tensor + Shape -> Tensor.
+            # Disallow mixing Tensor and Shape by default. If it makes sense in a given function
+            # to have both Tensor and Shape arguments, that might suggest that custom handling
+            # rather than relying on this decorator would make sense.
+            tensor_args = []
+            shape_args = []
+            # Don't need special handling for ShapeScalars because they get broadcast up into larger sizes
+            # as needed -- can treat them as a tensor or a shape whenever we need
+            shape_scalar_encountered = False
+            for arg_name, arg in all_args:
+                if arg_name in exclude:
+                    continue
+                if isinstance(arg, Shape):
+                    shape_args.append((arg_name, arg))
+                elif isinstance(arg, ShapeScalar):
+                    shape_scalar_encountered = True
+                elif isinstance(arg, Tensor):
+                    tensor_args.append((arg_name, arg))
+
+            if len(shape_args) and len(tensor_args):
+
+                def format_args(arg_info):
+                    return ", ".join(map(lambda name_value: f"{name_value[1]} (`{name_value[0]}`)", arg_info))
+
+                raise_error(
+                    "This operator expects all arguments to be tp.Tensor or all to be tp.Shape but was given arguments of mixed types."
+                    " Consider explicitly converting between these types using tp.Shape(value) or value.as_tensor()."
+                    f" Arguments that are tp.Shape: {format_args(shape_args)}."
+                    f" Arguments that are tp.Tensor: {format_args(tensor_args)}."
+                )
+
+            # Result is a shape scalar only if we can't broadcast it up to anything else
             TensorType = None
-            MaxType = max(
-                (type(arg) for arg_name, arg in all_args if arg_name not in exclude),
-                key=lambda typ: {Tensor: 2, Shape: 1, ShapeScalar: 0}.get(typ, -1),
-            )
-            if issubclass(MaxType, Tensor):
-                TensorType = MaxType
+            if shape_scalar_encountered and not len(tensor_args) and not len(shape_args):
+                TensorType = ShapeScalar
+            if len(tensor_args):
+                TensorType = Tensor
+            if len(shape_args):
+                TensorType = Shape
```

Comment:
I think this can be simplified to:
```py
types = {type(arg) for _, arg in all_args if isinstance(arg, Tensor)}
shape_scalar_encountered = ShapeScalar in types
types -= {ShapeScalar}

if len(types) > 1:
    raise_error(
        f"{func.__name__} expects tensor arguments to have matching class types, "
        f"but got mixed `tp.Tensor` and `tp.Shape` arguments",
        [
            "Consider explicitly converting using tp.Shape(tensor) or shape.as_tensor()\n"
            "Note: argument types were: " + ", ".join(f"{name}: {type(arg)}" for name, arg in all_args)
        ],
    )

TensorType = None
if types:
    TensorType = types.pop()
elif shape_scalar_encountered:
    TensorType = ShapeScalar
```

Reply:
Very Pythonic! I will do that.